### ISSUE TITLE
Storing structs to SSBOs break validation

### ISSUE THREAD

**Issue by Themaister** [KhronosGroup/glslang#304]
Referring to this issue:
https://github.com/KhronosGroup/SPIRV-Tools/issues/217

```
error: 196: OpStore Pointer <id> '10's type does not match Object <id> '11's type.
```

SPIR-V:
```
; SPIR-V
; Version: 1.0
; Generator: Khronos Glslang Reference Front End; 1
; Bound: 25
; Schema: 0
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %4 "main"
               OpExecutionMode %4 LocalSize 1 1 1
               OpSource ESSL 310
               OpName %4 "main"
               OpName %8 "MyStruct"
               OpMemberName %8 0 "foo"
               OpName %10 "bar"
               OpName %11 "MyStruct"
               OpMemberName %11 0 "foo"
               OpName %12 "SSBO0"
               OpMemberName %12 0 "foo_std430"
               OpName %14 ""
               OpName %20 "SSBO1"
               OpMemberName %20 0 "foo_std430_out"
               OpName %22 ""
               OpMemberDecorate %11 0 Offset 0
               OpMemberDecorate %12 0 Offset 0
               OpDecorate %12 BufferBlock
               OpDecorate %14 DescriptorSet 0
               OpDecorate %14 Binding 0
               OpMemberDecorate %20 0 Offset 0
               OpDecorate %20 BufferBlock
               OpDecorate %22 DescriptorSet 0
               OpDecorate %22 Binding 1
          %2 = OpTypeVoid
          %3 = OpTypeFunction %2
          %6 = OpTypeFloat 32
          %7 = OpTypeVector %6 4
          %8 = OpTypeStruct %7
          %9 = OpTypePointer Function %8
         %11 = OpTypeStruct %7
         %12 = OpTypeStruct %11
         %13 = OpTypePointer Uniform %12
         %14 = OpVariable %13 Uniform
         %15 = OpTypeInt 32 1
         %16 = OpConstant %15 0
         %17 = OpTypePointer Uniform %11
         %20 = OpTypeStruct %11
         %21 = OpTypePointer Uniform %20
         %22 = OpVariable %21 Uniform
          %4 = OpFunction %2 None %3
          %5 = OpLabel
         %10 = OpVariable %9 Function
         %18 = OpAccessChain %17 %14 %16
         %19 = OpLoad %11 %18
               OpStore %10 %19
         %23 = OpLoad %8 %10
         %24 = OpAccessChain %17 %22 %16
               OpStore %24 %23
               OpReturn
               OpFunctionEnd
```

Generated from this glslang output:
```
#version 310 es
layout(local_size_x = 1) in;

struct MyStruct
{
	vec4 foo;
};

layout(binding = 0, std430) buffer SSBO0
{
	MyStruct foo_std430;
};

layout(binding = 1, std430) buffer SSBO1
{
	MyStruct foo_std430_out;
};

void main()
{
	MyStruct bar = foo_std430;
	foo_std430_out = bar;
}
```

**Comment by johnkslang** [comment]
I think what's missing here is member-by-member copy, when the two structs are the same type at the GLSL level, but not at the SPIR-V level.

This can happen because, in the general case, the struct has two different layouts, one for each block it is in.  So, it might need two sets of decorations.  (The scheme to reuse types is close, but not perfect, so sometimes there will be two types when there didn't need to be.  However, the general case has to be handled anyway: sometimes there will be two different types.)

**Comment by godlikepanos** [comment]
Is there any progress on this bug? It seems like a quite common case and if the struct is big it gets kind of difficult to workaround.

**Comment by johnkslang** [comment]
I'm opening a Khronos issue to discuss how to resolve this.  It does seem a common case to want to deal with more efficiently, but that will require a change to SPIR-V.

**Comment by johnkslang** [comment]
This specific part of the issue is now fixed in commit 4bf7155, by doing recursive/flattening copies of aliased structure types.  This commit also mostly address issue #307, so will close this one and reopen that one.
